lustre/ws.lua

```diff
@@ -0,0 +1,351 @@
+        if (bytes:len() - data_idx + 1) > self.config._max_frame_size then
+            header.set_fin(false)
+        end
+        payload = string.sub(bytes, data_idx, data_idx + self.config._max_frame_size)
+        header:set_opcode(OpCode.binary())
+        header:set_length(#payload)
+        local frame = Frame.from_parts(
+            header,
+            payload
+        )
+        frame:set_mask() --TODO handle client vs server
+        local sent_bytes, err = self.socket:send(frame:encode())
+        --TODO send frame over cosock channel to be sent on socket from receive loop
+        if not sent_bytes then
+            return err
+        end
+        data_idx = data_idx + sent_bytes
+        frames_sent = frames_sent + 1
+    end
+end
+
+---@param message Message
+---@return err string|nil
+function WebSocket:send(message)
+    return "not implemented"
+end
+
+---@return success number 1 if handshake was successful
+---@return err string|nil
+function WebSocket:connect(host, port)
+    if not self.is_client then --todo use metatables to enforce this
+        return nil, "only a client can connect"
+    end
+    if not host or not port then
+        return nil, "missing host or port"
+    end
+
+    local r, err = self.socket:connect(host, port)
+    if not r then
+        return nil, "socket connect failure: "..err
+    end
+
+    --Do handshake
+    local req = Request.new('GET', self.url, self.socket)
+    req:add_header('Connection', 'Upgrade')
+    req:add_header('Upgrade', 'websocket')
+    req:add_header('User-Agent', "lua-lustre")
+    req:add_header('Sec-Websocket-Version', 13)
+    req:add_header('Sec-Websocket-Key', self.handshake_key)
+    req:add_header("Host", string.format("%s:%d", host, port))
+    for _, prot in ipairs(self.config.protocols) do
+        --TODO I think luncheon should be able to handle multiple values, but
+        -- it currently only sends the last value added
+        req:add_header("Sec-Websocket-Protocol", prot)
+    end
+    local s, err = req:send()
+    if not s then
+        return "handshake request failure: "..err
+    end
+    local res, err = Response.tcp_source(self.socket)
+    if not res then
+        return nil, "Upgrade response failure: "..err
+    end
+    local handshake = Handshake.client(self.handshake_key, {}, {})
+    local success, err = handshake:validate_accept(res)
+    if not success then
+        return nil, 'invalid handshake: '..err
+    end
+    cosock.spawn(function() self:receive_loop() end, 'Client receive loop')
+    return 1
+end
+
+function WebSocket:accept()
+end
+
+---@param close_code CloseCode
+---@param reason string
+---@return err string|nil
+function WebSocket:close(close_code, reason)
+    local close_frame = Frame.close(close_code, reason):set_mask()
+    local sent, err = self.socket:send(close_frame:encode())
+    if not sent then return nil, err end
+    print("SENT CLOSE FRAME: \n", utils.table_string(close_frame), "\n\n")
```

Comment:
This should probably be using `log.debug` over `print` which would require including `logface` in the dependencies